TLSharp.Core/Network/TcpTransport.cs

```diff
@@ -10,103 +12,64 @@ namespace TLSharp.Core.Network
 
     public class TcpTransport : IDisposable
     {
-        private readonly TcpClient tcpClient;
-        private readonly NetworkStream stream;
+        private readonly TLClient tcpClient;
+        private readonly BlockingCollection<TcpMessage> ReceievedMessage = new BlockingCollection<TcpMessage>();
+        //private readonly NetworkStream stream;
         private int sendCounter = 0;
 
         public TcpTransport(string address, int port, TcpClientConnectionHandler handler = null)
         {
-            if (handler == null)
-            {
-                var ipAddress = IPAddress.Parse(address);
-                var endpoint = new IPEndPoint(ipAddress, port);
+            //if (handler == null)
+            //{
+            var ipAddress = IPAddress.Parse(address);
+            tcpClient = new TLClient(ipAddress, port, ReceiveMessage);
+            tcpClient.ConnectAsync();
+            //}
+            //else
+            //    tcpClient = handler(address, port);
+        }
 
-                tcpClient = new TcpClient(ipAddress.AddressFamily);
-                tcpClient.Connect(endpoint);
-            }
-            else
-                tcpClient = handler(address, port);
+        public void ReceiveMessage(TcpMessage message)
+        {
+            ReceievedMessage.Add(message);
```

Comment:
Old Code Compatibility. It's here to make sync receive function work.
In the final version we should read the message deserialize it find the type and call the handler for that type.
for example if we get a TLMessage we should search a dictionary<type,delegate> to find the handler function.